### PR TITLE
[export] Add support for explicit sharding.

### PR DESCRIPTION
**Pull request by gnecula**
[export] Add support for explicit sharding.

This is a significant addition to exporting JAX functions.

Previously, the `Exported` objects and their serializations did not store the meshes, in the idea that one may want to
use the `Exported` object with another concrete mesh, as long as the number of devices match.

Starting with this PR, we added new fields to `Exported` to contain NamedShardings (including abstract meshes, and partition specs). This is needed to suport explicit sharding in JAX.

One additional advantage of this change is that we do not need to rely on Shardy to store the mesh in the serialized
HLO. (Those Shardy-stored meshes do not support multiple meshes for different inputs/ouputs, and currently do not
include the `axis_type`).

There is one regression: The function `in_shardings_jax(new_mesh)` previously cared only for `new_mesh` to have the same number of devices. Now we check that the `new_mesh` has the same abstract mesh as the one that was used for serialization.

I have verified manually forward and backward compatibility using the not-yet-submitted compatibility tests in #33758.

#jax-fixit

### CODE REVIEW

**Comment by gnecula** [comment]
@yashk2810 some questions about the plan.

I plan to export and serialize only `NamedSharding` with `AbstractMesh`. Do I need to worry about other kind of shardings? Perhaps `UNSPECIFIED`?

For the `AbstractMesh` I want to make the mesh names not binding (the caller should not be bound by the exact mesh names used when exporting). One way to do this, is to actually include the names, both in the `in_avals` and `out_avals` but when calling the exported be ready to map the exported axis names to the axis names in the importing program.

How to do the mapping? We can require that the importing mesh has the same axis sizes and axis types, and then we can match the names to compute an axis name mapping.

Are there places in the lowered HLO where the axis names appear?

What to do with devices (the logical device ids in the `NamedSharding` and the `abstract_device` in the `AbstractMesh`)? Exporting works cross-platform and multi-platform, so it is designed to allow calling with different devices and even different platforms than when exporting. We even have tests for this, e.g., to run the exported code with a different device assignment. One solution is to not save the AbstactDevice nor the logical device ids. I assume that in this case we automatically use the device assignment from the importing program.

**Comment by yashk2810** [comment]
> Do I need to worry about other kind of shardings? Perhaps UNSPECIFIED?

No, only NamedSharding. Is there flexibility in changing NamedSharding's API once you export it? Cause I plan to rename it in fixit week. Also there might be changes coming to AbstractMesh, PartitionSpec, etc in the future and I want the flexibility to change those APIs whenever I want.

> Are there places in the lowered HLO where the axis names appear?

Yes, with shardy the axis names are baked into the stableHLO. But shardy can handle multiple meshes so it should be fine.

> We can require that the importing mesh has the same axis sizes and axis types, and then we can match the names to compute an axis name mapping.

Yes, this sounds good to me. Can you please add a test for this so I can look at it?

> One solution is to not save the AbstactDevice nor the logical device ids

Yup! No need to serialize logical_device_ids (this is not concrete devices but it has no effect with ShiT). 

For AbstractDevice, you can leave it for now. As you said, export allows multi-platform so no need to serialize AbstractDevice too.

File: jax/_src/export/serialization.py [hunk 1]
```diff
@@ -397,6 +435,153 @@ def register_dtype_kind(dtype: Any, kind: int):
...
+    axis_types = None
+  else:
+    axis_types = tuple(_axis_type_from_enum[ser_mesh.AxisTypes(i)]
+                       for i in range(ser_mesh.AxisTypesLength()))
+  return mesh.AbstractMesh(axis_sizes, axis_names, axis_types)
+
+
+def _serialize_partition_spec_one_axis(builder: flatbuffers.Builder,
+                                       spec: str | tuple[str, ...] | None) -> int:
+  if spec is None:
```
Comment (yashk2810) [inline jax/_src/export/serialization.py:476]: this can be 1 line :)

```
axes = () if spec is None else ((spec,) if isinstance(spec, str) else spec)
```

File: jax/_src/core.py [hunk 2]
```diff
@@ -2186,7 +2186,7 @@ def get_sharding(sharding, shape):
     raise ValueError("Mesh of an aval must be an AbstractMesh. "
                      f"Got {out_s.mesh} of type {type(out_s.mesh)}")
   _check_divisibility(out_s, shape)
-  assert out_s.memory_kind is None
+  # assert out_s.memory_kind is None  # TODO(necula): discuss this with Yash
```
Comment (yashk2810) [inline jax/_src/core.py]: Why this? This is an important check

Comment (gnecula) [reply jax/_src/core.py]: The previous assertion fails when I try to construct ShapedArray on deserialization, if I pass a sharding with memory_kind filled in. Why is that an invalid use of the constructor? We do end up with ShapedArray with non-None sharding.memory_kind at some point, right?

Comment (gnecula) [reply jax/_src/core.py]: There is another problem with this constructor: if we don't pass a sharding, but we do pass a memory_space, the memory kind is not stored in the NamedSharding that is constructed. Then we end up with a mismatch with the aval having the specified memory_space and the sharding having no memory kind.

File: jax/_src/export/serialization.py (hunk 1 above)
Comment (gnecula) [reply jax/_src/export/serialization.py:476]: I changed to a 3-line version

File: jax/_src/core.py (hunk 2 above)
Comment (yashk2810) [reply jax/_src/core.py]: > We do end up with ShapedArray with non-None sharding.memory_kind at some point, right?

Nope, never! Hence the check.

aval tracks memory spaces in `aval.memory_space` not in `aval.sharding.memory_kind`.

> Then we end up with a mismatch with the aval having the specified memory_space and the sharding having no memory kind

What kind of mismatch? Just don't read the memory_kind of an aval sharding ever. Then there is no mismatch.

Comment (gnecula) [reply jax/_src/core.py]: Discussed offline. The conclusion is that when a NamedSharing is used in ShapedArray.sharding, then its `memory_kind` should be `None`, and the information is tracked in `ShapedArray.memory_space`. We achieve this by populating the `sharding` field using `core.update_aval_with_sharding`.

File: jax/_src/named_sharding.py [hunk 3]
```diff
@@ -108,6 +108,9 @@ class NamedSharding(JSharding.Sharding):
   Args:
     mesh: A :class:`jax.sharding.Mesh` object.
     spec: A :class:`jax.sharding.PartitionSpec` object.
+    memory_kind: A string indicating the memory kind of the sharding. Note that
```
Comment (yashk2810) [inline jax/_src/named_sharding.py]: can you not mention the ShapedArray thing? That seems unrelated to NamedSharding docstring.

Comment (gnecula) [reply jax/_src/named_sharding.py]: Ok, I moved this to the ShapedArray docstring.

File: jax/_src/core.py [hunk 4]
```diff
@@ -2238,6 +2238,7 @@ def __init__(self, shape, dtype, weak_type=False, *, sharding=None,
     self.shape = canonicalize_shape(shape)
     self.dtype = _dtype_object(dtype)
     self.weak_type = weak_type
+    # The sharding.memory_kind should not be relied upon, used memory_space.
```
Comment (yashk2810) [inline jax/_src/core.py]: Typo?

s/used/use/g?

**Review by yashk2810 — APPROVED** [review]
Lgtm.. thanks!

File: jax/_src/core.py (hunk 4 above)
Comment (yashk2810) [reply jax/_src/core.py]: Or maybe..

aval.sharding.memory_kind is always None. Use aval.memory_space.

Comment (gnecula) [reply jax/_src/core.py]: done
